Source/Tests/LinuxSyscalls/Syscalls.h

```diff
@@ -203,6 +218,97 @@ class SyscallHandler : public FEXCore::HLE::SyscallHandler {
   #endif
 
   std::unique_ptr<FEX::HLE::MemAllocator> Alloc32Handler{};
+  
+  ///// VMA (Virtual Memory Area) tracking /////
+
+  FEXCore::HLE::AOTIRCacheEntryLookupResult LookupAOTIRCacheEntry(uint64_t GuestAddr) final override;
+
+
+  struct MRID {
+    uint64_t dev; /* kernel only uses lower 32-bits of dev_t */
+    uint64_t id;
+
+    bool operator<(const MRID& other) const {
+      return std::tie(dev, id) < std::tie(other.dev, other.id);
+    }
+  };
+
+  // We use bits 63:32 for our own custom dev tracking
+  static constexpr const uint64_t MRID_PRIVATE = 0x1'0000'0000ULL;
+  static constexpr const uint64_t MRID_ANON = 0x2'0000'0000ULL;
+  static constexpr const uint64_t MRID_SHM =  0x3'0000'0000ULL;
+
+  struct VMAEntry;
+
+  struct MappedResource {
+    typedef std::map<MRID, MappedResource> ContainerType;
+
+    FEXCore::IR::AOTIRCacheEntry *AOTIRCacheEntry;
+    VMAEntry *FirstVMA;
+    uint64_t Length; // 0 if not fixed size
+    ContainerType::iterator Iterator;
+  };
+
+  struct VMAProt {
+    bool Writable: 1;
+    bool operator==(const VMAProt& other) const {
+      return other.Writable == Writable;
+    }
+  };
+
+  struct VMAOptions {
+    VMAOptions(int Writable, const int Shared) : Prot {Writable != 0}, Flags {Shared != 0} { }
+
+    VMAProt Prot;
+
+    const struct {
+      bool Shared: 1;
+    } Flags;
+  };
+
+  struct VMAEntry {
+    MappedResource* Resource;
+    VMAEntry* ResourcePrevVMA;
+    VMAEntry* ResourceNextVMA;
+
+    uint64_t Base;
+    uint64_t Offset;
+    uint64_t Length;
+
+    VMAOptions Options;
+  };
+
+  struct VMATracking {
+    using VMAEntry = SyscallHandler::VMAEntry;
+    std::shared_mutex Mutex{};
+    
+    // Memory ranges indexed by page aligned starting address
+    typedef std::map<uint64_t, VMAEntry> VMAsType;
+    VMAsType VMAs;
+    MappedResource::ContainerType MappedResources;
+
+    // Mutex must be at least shared_locked before calling
```

Comment:
Would it be reasonable to validate these assumptions in debug mode using `shared_mutex::try_lock_shared()`?